content/blog/2023/2023-12-12-enum-vs-seealed-classes-in-kotlin.md

```diff
@@ -0,0 +1,52 @@
+---
+title: "Sealed classes vs Enum classes in Kotlin"
+categories: ["Kotlin"]
+date: 2023-12-12 00:00:00 +1100
+authors: [ezra]
+excerpt: "In this tutorial, we'll discuss differences between sealed classes and enum classes in Kotlin"
+image: images/stock/0104-on-off-1200x628-branded.jpg
+url: sealed vs enum classses in kotlin
+---
+
+Kotlin, a modern programming language brings a range of features to enhance expressiveness and conciseness. Two key constructs in Kotlin that contribute to its versatility are sealed classes and enum classes. In this blog post, we'll delve into the characteristics, use cases and differences between sealed classes and enum classes.
+
+## Sealed Classes:
+Sealed classes in Kotlin offer a powerful tool for defining restricted class hierarchies. When a class is marked as sealed, it means that the class hierarchy is finite and every subclass must be declared within the same file. This restriction allows the compiler to perform exhaustive checks when used in a when expression, ensuring that all possible subclasses are covered.
+
+A typical use case for sealed classes is modeling hierarchical data structures, such as expressions in a compiler or states in a finite state machine. By sealing the class hierarchy, developers can guarantee that they handle all possible cases, making the code more robust and less prone to bugs.
+
+Here's an example of a sealed class representing mathematical expressions:
+
+```kotlin
+sealed class MathExpression {
+    data class Value(val value: Double) : MathExpression()
+    data class Addition(val left: MathExpression, val right: MathExpression) : MathExpression()
+    data class Subtraction(val left: MathExpression, val right: MathExpression) : MathExpression()
+    object Undefined : MathExpression()
+}
+
+```
+
+### Enum Classes
+Enum classes, short for enumerated classes, are another feature that Kotlin inherits from Java but enhances significantly. Enum classes allow developers to define a fixed set of values, each of which is an instance of the enum class. Enums are particularly useful when modeling a closed set of related constants.
+
+Let's consider an example of an enum class representing the days of the week:
+```kotlin
+enum class DayOfWeek {
+    SUNDAY, MONDAY, TUESDAY, WEDNESDAY, THURSDAY, FRIDAY, SATURDAY
+}
+
+```
+Unlike sealed classes, enum classes can't have subclasses, and the set of values is predetermined at compile time. This makes enum classes suitable for scenarios where a predefined set of options is expected, like representing days, months, or error states.
```

Comment:
"and the set of values is predetermined at compile time" ... isn't the set of values (or rather subclasses) of sealed classes also predetermined at compile time?